Makefile

```diff
@@ -32,18 +32,43 @@ SUB := $(addsuffix .sub, $(ALL))
 build: build.sub
 build-linux: build-linux.sub
 build-darwin: build-darwin.sub
-clean: clean.sub
 docker: docker.sub
 lint: lint.sub
 push: push.sub
 test: test.sub
 
+clean: clean.sub
+	rm -f .dockerInit
+	rm -f .buildImage
+	docker rmi -f buildImage > /dev/null 2>&1 || true
+
+# Use this target when you want to build everything using docker containers.
+# Good for cases when you don't have the tools installed (like glide or go).
+docker-all: .buildImage
+	docker run --rm -ti \
+	  -v $(PWD):/go/src/github.com/kubernetes-incubator/service-catalog \
+	  buildimage \
+	  make .dockerInit all
+
+# .dockerInit tells us if our vendor stuff if out of date or not.
+# And if so we'll run init under our docker build.  For non-docker builds
+# it is assumed you'll run "make init" manually.
+.dockerInit: glide.yaml
+	make init
+	echo > .dockerInit
+
+# .buildImage is used to know when the docker image ("buildimage") is out
+# of date with the Dockerfile.
+.buildImage: hack/Dockerfile
+	docker build -t buildimage - < hack/Dockerfile
+	echo > .buildImage
```

Comment:
Ah... I see. Two follow up comments to that.

1. My preference would be that we _do_ rebuild the buildimage every time. Docker caching will make this quick, so it's really not much of an issue at all. And this will have the benefit of never missing a _change_ to the buildimage itself.

1. If somehow people don't agree with the above, then at minimum, the `clean` target should be added to `.travis.yml` so that CI will always rebuild the buildimage before proceeding. (but I still feel strongly that we should prefer to rebuild it each time).

Avoiding running glide each time, however, does make sense to me, as that's a more time-consuming operation.